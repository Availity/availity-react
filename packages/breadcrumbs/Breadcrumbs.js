import React from 'react';
import PropTypes from 'prop-types';
import { Breadcrumb, BreadcrumbItem } from 'reactstrap';

const Breadcrumbs = ({
  crumbs,
  active,
  emptyState,
  children,
  linkTag: LinkTag,
  homeUrl,
  ...rest
}) => {
  const renderBreadCrumb = crumb => {
    // default breadcrumbitem render
    let breadCrumbItemChildren = <span>{emptyState}</span>;
    // render static links
    if (crumb.name && crumb.url) {
      breadCrumbItemChildren = (
        <LinkTag aria-label={crumb.name} href={crumb.url}>
          {crumb.name}
        </LinkTag>
      );
    }
    return (
      <BreadcrumbItem key={crumb.name + crumb.url}>
        {breadCrumbItemChildren}
      </BreadcrumbItem>
    );
  };

  return (
    <Breadcrumb {...rest}>
      <BreadcrumbItem>
        <LinkTag aria-label="Home" href={homeUrl}>
          Home
        </LinkTag>
      </BreadcrumbItem>
      {crumbs && crumbs.length > 0 && crumbs.map(renderBreadCrumb)}
      {children}
      <BreadcrumbItem active>{active || emptyState}</BreadcrumbItem>
    </Breadcrumb>
  );
};

Breadcrumbs.propTypes = {
  crumbs: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      url: PropTypes.string,
    })
  ),
  linkTag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  active: PropTypes.string.isRequired,
  emptyState: PropTypes.string,
  children: PropTypes.node,
  homeUrl: PropTypes.string,
};

Breadcrumbs.defaultProps = {
  emptyState: 'â€¦',
  homeUrl: '/public/apps/dashboard',
  linkTag: 'a',
};

export default Breadcrumbs;
