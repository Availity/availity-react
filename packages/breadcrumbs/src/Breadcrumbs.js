import React from 'react';
import PropTypes from 'prop-types';
import { Breadcrumb, BreadcrumbItem } from 'reactstrap';

const Breadcrumbs = ({ crumbs, active, emptyState, children, linkTag: LinkTag, homeUrl, ...rest }) => {
  const renderBreadCrumb = (crumb) => {
    // default breadcrumbitem render
    let breadCrumbItemChildren = <span>{emptyState}</span>;
    // render static links
    if (crumb.name && crumb.url) {
      breadCrumbItemChildren = (
        <LinkTag aria-label={crumb.name} href={crumb.url}>
          {crumb.name}
        </LinkTag>
      );
    }
    return <BreadcrumbItem key={crumb.name + crumb.url}>{breadCrumbItemChildren}</BreadcrumbItem>;
  };

  return (
    <Breadcrumb {...rest}>
      <BreadcrumbItem>
        <LinkTag aria-label="Home" href={homeUrl}>
          Home
        </LinkTag>
      </BreadcrumbItem>
      {crumbs && crumbs.length > 0 && crumbs.map((crumb) => renderBreadCrumb(crumb))}
      {children}
      <BreadcrumbItem active>{active || emptyState}</BreadcrumbItem>
    </Breadcrumb>
  );
};

Breadcrumbs.propTypes = {
  /** The name of the active page (the page the user is currently on). */
  active: PropTypes.string.isRequired,
  /** The ancestor pages. Objects in array contain `name` (String) and `url` (String) properties. */
  crumbs: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      url: PropTypes.string,
    })
  ),
  /** Custom link tag for the links. */
  linkTag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /** The value to display when the active page or an ancestor does not have a value. */
  emptyState: PropTypes.string,
  /** The children must be a reactstrap `BreadcrumbItem` components. */
  children: PropTypes.node,
  /** Url for the Home route. */
  homeUrl: PropTypes.string,
};

Breadcrumbs.defaultProps = {
  emptyState: 'â€¦',
  homeUrl: '/static/web/onb/onboarding-ui-apps/dashboard-ui/',
  linkTag: 'a',
};

export default Breadcrumbs;
