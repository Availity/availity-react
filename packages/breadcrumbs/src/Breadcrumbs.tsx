import React from 'react';
import { Breadcrumb, BreadcrumbItem, BreadcrumbProps } from 'reactstrap';

type Crumb = {
  name: string;
  url: string;
};

type Props = {
  active: string;
  children?: React.ReactNode;
  crumbs?: Crumb[];
  emptyState?: string;
  homeUrl?: string;
  linkTag?: React.ElementType;
} & BreadcrumbProps;

const Breadcrumbs = ({
  crumbs,
  active,
  emptyState = '...',
  children,
  linkTag: LinkTag = 'a',
  homeUrl = '/public/apps/dashboard',
  ...props
}: Props): JSX.Element => {
  const renderCrumb = ({ name, url }: Crumb) => {
    let item = <span>{emptyState}</span>;

    if (name && url) {
      item = (
        <LinkTag aria-label={name} href={url}>
          {name}
        </LinkTag>
      );
    }

    return <BreadcrumbItem key={name + url}>{item}</BreadcrumbItem>;
  };

  return (
    <Breadcrumb {...props}>
      <BreadcrumbItem>
        <LinkTag aria-label="Home" href={homeUrl}>
          Home
        </LinkTag>
      </BreadcrumbItem>
      {crumbs?.map((crumb) => renderCrumb(crumb)) || null}
      {children}
      <BreadcrumbItem active>{active || emptyState}</BreadcrumbItem>
    </Breadcrumb>
  );
};

export default Breadcrumbs;
