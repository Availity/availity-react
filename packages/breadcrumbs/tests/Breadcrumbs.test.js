import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { NavLink } from 'reactstrap';
import Breadcrumbs from '..';

afterEach(cleanup);

describe('Breadcrumbs', () => {
  test('should render', () => {
    const { container } = render(<Breadcrumbs active="" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render child as current page', () => {
    const { container } = render(<Breadcrumbs active="Payer Space" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render crumbs', () => {
    const crumbs = [
      {
        name: 'my grand parent page',
        url: '/parent-page',
      },
      {
        name: 'my parent page',
        url: '/grand-parent-page',
      },
    ];
    const { container } = render(
      <Breadcrumbs crumbs={crumbs} active="Payer Space" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render crumbs with missing values', () => {
    const crumbs = [
      {
        name: 'my grand parent page',
        url: '/parent-page',
      },
      {
        name: '',
        url: '',
      },
    ];
    const { container } = render(
      <Breadcrumbs crumbs={crumbs} active="Payer Space" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render custom ellipse', () => {
    const crumbs = [
      {
        name: '',
        url: '',
      },
    ];
    const { container } = render(
      <Breadcrumbs crumbs={crumbs} emptyState="???" active="Payer Space" />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render custom home url', () => {
    const { getByText } = render(
      <Breadcrumbs homeUrl="/go-home" active="Payer Space" />
    );

    const homeBtn = getByText('Home');

    expect(homeBtn.getAttribute('href')).toBe('/go-home');
  });

  test('should render custom link tag', () => {
    const { getByText } = render(
      <Breadcrumbs homeUrl="/go-home" active="Payer Space" linkTag={NavLink} />
    );

    const homeBtn = getByText('Home');

    expect(homeBtn.className).toBe('nav-link');
  });
});
