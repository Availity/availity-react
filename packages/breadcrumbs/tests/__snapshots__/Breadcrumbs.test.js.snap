// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should render 1`] = `
<nav
  aria-label="breadcrumb"
  class=""
>
  <ol
    class="breadcrumb"
  >
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="Home"
        href="/public/apps/dashboard"
      >
        Home
      </a>
    </li>
    <li
      aria-current="page"
      class="active breadcrumb-item"
    >
      …
    </li>
  </ol>
</nav>
`;

exports[`Breadcrumbs should render child as current page 1`] = `
<nav
  aria-label="breadcrumb"
  class=""
>
  <ol
    class="breadcrumb"
  >
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="Home"
        href="/public/apps/dashboard"
      >
        Home
      </a>
    </li>
    <li
      aria-current="page"
      class="active breadcrumb-item"
    >
      Payer Space
    </li>
  </ol>
</nav>
`;

exports[`Breadcrumbs should render crumbs 1`] = `
<nav
  aria-label="breadcrumb"
  class=""
>
  <ol
    class="breadcrumb"
  >
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="Home"
        href="/public/apps/dashboard"
      >
        Home
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="my grand parent page"
        href="/parent-page"
      >
        my grand parent page
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="my parent page"
        href="/grand-parent-page"
      >
        my parent page
      </a>
    </li>
    <li
      aria-current="page"
      class="active breadcrumb-item"
    >
      Payer Space
    </li>
  </ol>
</nav>
`;

exports[`Breadcrumbs should render crumbs with missing values 1`] = `
<nav
  aria-label="breadcrumb"
  class=""
>
  <ol
    class="breadcrumb"
  >
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="Home"
        href="/public/apps/dashboard"
      >
        Home
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="my grand parent page"
        href="/parent-page"
      >
        my grand parent page
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <span>
        …
      </span>
    </li>
    <li
      aria-current="page"
      class="active breadcrumb-item"
    >
      Payer Space
    </li>
  </ol>
</nav>
`;

exports[`Breadcrumbs should render custom ellipse 1`] = `
<nav
  aria-label="breadcrumb"
  class=""
>
  <ol
    class="breadcrumb"
  >
    <li
      class="breadcrumb-item"
    >
      <a
        aria-label="Home"
        href="/public/apps/dashboard"
      >
        Home
      </a>
    </li>
    <li
      class="breadcrumb-item"
    >
      <span>
        ???
      </span>
    </li>
    <li
      aria-current="page"
      class="active breadcrumb-item"
    >
      Payer Space
    </li>
  </ol>
</nav>
`;
